builtin/providers/ultradns/resource_ultradns_dirpool.go

```diff
@@ -0,0 +1,612 @@
+package ultradns
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/Ensighten/udnssdk"
+	"github.com/fatih/structs"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/mitchellh/mapstructure"
+)
+
+func resourceUltradnsDirpool() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceUltradnsDirpoolCreate,
+		Read:   resourceUltradnsDirpoolRead,
+		Update: resourceUltradnsDirpoolUpdate,
+		Delete: resourceUltradnsDirpoolDelete,
+
+		Schema: map[string]*schema.Schema{
+			// Required
+			"zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				// 0-255 char
+			},
+			"rdata": &schema.Schema{
+				Type:     schema.TypeSet,
+				Set:      hashRdatas,
+				Required: true,
+				// Valid: len(rdataInfo) == len(rdata)
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						// Required
+						"host": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"all_non_configured": &schema.Schema{
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+						},
+						"geo_info": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": &schema.Schema{
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+									"is_account_level": &schema.Schema{
+										Type:     schema.TypeBool,
+										Optional: true,
+										Default:  false,
+									},
+									"codes": &schema.Schema{
+										Type:     schema.TypeSet,
+										Optional: true,
+										Elem:     &schema.Schema{Type: schema.TypeString},
+										Set:      schema.HashString,
+									},
+								},
+							},
+						},
+						"ip_info": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": &schema.Schema{
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+									"is_account_level": &schema.Schema{
+										Type:     schema.TypeBool,
+										Optional: true,
+										Default:  false,
+									},
+									"ips": &schema.Schema{
+										Type:     schema.TypeSet,
```

Comment:
it may be worth splitting this block out - it's 4 levels deep in the schema and it could be "interesting" in getting / setting the values